# 개요
- 참고 자료를 보셔도 좋지만, 코드 이해뿐만 아니라 데이터 전처리 처리에 대한 순서 / 법칙을 이해하면서 작성합시다!
- 코드에 대한 다양한 도전을 합시다 :) 
- 원본 데이터 : https://www.kaggle.com/datasets/fedesoriano/stroke-prediction-dataset/data

## 문제 1: 데이터 불러오기와 크기 확인
뇌졸중 예측 데이터셋 CSV 파일을 pandas를 사용하여 불러오고, 데이터프레임의 크기(행과 열의 개수)를 출력하는 코드를 작성하세요. (힌트: pd.read_csv 함수를 사용합니다.)

In [7]:
import pandas as pd
stroke_df = pd.read_csv('../data/stroke-data.csv')
stroke_df.shape

(5110, 12)

## 문제 2: 데이터 미리보기 (앞부분 확인)
데이터가 잘 불러와졌는지 확인하기 위해 데이터프레임의 처음 5행을 출력하는 코드를 작성하세요. 이때 데이터의 각 열 이름도 함께 확인해보세요. 그리고 각 열의 의미를 적어보세요!  
그리고 특정 열이 어떤 값이 있는지 확인해보세요! (특정열.unique() 사용)

In [8]:
stroke_df.head() # 방법1

display(stroke_df)  # 방법2

id  gender   age  hypertension  heart_disease ever_married  \
0      9046    Male  67.0             0              1          Yes   
1     51676  Female  61.0             0              0          Yes   
2     31112    Male  80.0             0              1          Yes   
3     60182  Female  49.0             0              0          Yes   
4      1665  Female  79.0             1              0          Yes   
...     ...     ...   ...           ...            ...          ...   
5105  18234  Female  80.0             1              0          Yes   
5106  44873  Female  81.0             0              0          Yes   
5107  19723  Female  35.0             0              0          Yes   
5108  37544    Male  51.0             0              0          Yes   
5109  44679  Female  44.0             0              0          Yes   

          work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0           Private          Urban             228.69  36.6  formerly smoked   
1     Self-employed          Rural             202.21   NaN     never smoked   
2           Private          Rural             105.92  32.5     never smoked   
3           Private          Urban             171.23  34.4           smokes   
4     Self-employed          Rural             174.12  24.0     never smoked   
...             ...            ...                ...   ...              ...   
5105        Private          Urban              83.75   NaN     never smoked   
5106  Self-employed          Urban             125.20  40.0     never smoked   
5107  Self-employed          Rural              82.99  30.6     never smoked   
5108        Private          Rural             166.29  25.6  formerly smoked   
5109       Govt_job          Urban              85.28  26.2          Unknown   

      stroke  
0          1  
1          1  
2          1  
3          1  
4          1  
...      ...  
5105       0  
5106       0  
5107       0  
5108       0  
5109       0  

[5110 rows x 12 columns]

In [29]:
stroke_df['avg_glucose_level'].unique()

array([228.69, 202.21, 105.92, ...,  82.99, 166.29,  85.28], shape=(3979,))

## 문제 3: 데이터 정보 확인 (데이터 타입과 결측값)
데이터프레임의 각 열에 대한 요약 정보를 출력하는 코드를 작성하세요. 이를 통해 각 열의 데이터 타입(dtype)과 누락된 값(null/결측값) 개수를 확인해보세요. (힌트: DataFrame.info() 메서드를 사용합니다.)

In [33]:
stroke_df['bmi'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 5110 entries, 0 to 5109
Series name: bmi
Non-Null Count  Dtype  
--------------  -----  
4909 non-null   float64
dtypes: float64(1)
memory usage: 40.0 KB


## 문제 3-1: 데이터 정보 확인 (특정 컬럼 값 분포 확인)
hypertension, ever_married, avg_glucose_level, stroke 열의 값 분포를 확인하세요! (힌트: value_counts() 메서드를 사용합니다.)

In [37]:
stroke_df['hypertension'].value_counts()

hypertension
0    4612
1     498
Name: count, dtype: int64

## 문제 4: 결측값 개수 확인
각 열별로 결측값(null)이 몇 개인지 합계로 확인하려 합니다. 이를 계산하는 코드를 작성하세요. (힌트: isnull() 또는 isna()와 sum() 함수를 활용합니다.)

In [9]:
#stroke_df['id'].isnull().sum()
stroke_df['bmi'].isna().sum()

np.int64(201)

## 문제 5: id 열 삭제하기
id 열은 각 환자를 식별하기 위한 고유 식별자일 뿐, 예측 모델링에는 필요하지 않은 정보입니다. 데이터프레임에서 id 열을 제거하는 코드를 작성하세요.

In [45]:
print(stroke_df.drop(columns=['id']))

      gender   age  hypertension  heart_disease ever_married      work_type  \
0       Male  67.0             0              1          Yes        Private   
1     Female  61.0             0              0          Yes  Self-employed   
2       Male  80.0             0              1          Yes        Private   
3     Female  49.0             0              0          Yes        Private   
4     Female  79.0             1              0          Yes  Self-employed   
...      ...   ...           ...            ...          ...            ...   
5105  Female  80.0             1              0          Yes        Private   
5106  Female  81.0             0              0          Yes  Self-employed   
5107  Female  35.0             0              0          Yes  Self-employed   
5108    Male  51.0             0              0          Yes        Private   
5109  Female  44.0             0              0          Yes       Govt_job   

     Residence_type  avg_glucose_level   bmi   smok

## 문제 6: BMI 결측값 처리 (중앙값 대체)
bmi 열에 존재하는 201개의 결측값을 어떻게 처리할지 결정해야 합니다. 이번에는 중앙값(median)을 사용하여 이 결측치를 채우려고 합니다. bmi 열의 중앙값을 계산하고, 그 값으로 결측값들을 채우는 코드를 작성하세요.

In [55]:
stroke_df['bmi'].isnull().sum()

stroke_df['bmi'].describe()

stroke_df['bmi'].median()

print(stroke_df['bmi'])
print(stroke_df['bmi'].isnull().sum())
print("------------------------------------------")
print(stroke_df['bmi'].fillna(stroke_df['bmi'].median()))
print(stroke_df['bmi'].fillna(stroke_df['bmi'].median()).isnull().sum())

0       36.6
1        NaN
2       32.5
3       34.4
4       24.0
        ... 
5105     NaN
5106    40.0
5107    30.6
5108    25.6
5109    26.2
Name: bmi, Length: 5110, dtype: float64
201
------------------------------------------
0       36.6
1       28.1
2       32.5
3       34.4
4       24.0
        ... 
5105    28.1
5106    40.0
5107    30.6
5108    25.6
5109    26.2
Name: bmi, Length: 5110, dtype: float64
0


## 문제 7: 행 삭제(drop) 연습
bmi 값이 결측치(NaN) 이거나, bmi < 10 또는 bmi > 80인 행은 잘못된 데이터라고 가정합시다.  
이런 행들을 제거한 새로운 데이터프레임 df_clean을 만드세요. (원본 df는 그대로 둡니다.)  
제거 전/후 행 개수를 비교해서 출력해보세요.  

In [ ]:
# 내가 작성한 코드
stroke_df.dropna(subset=['bmi'])

stroke_df.drop(stroke_df['bmi']<10 or stroke_df['bmi']>80)

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

In [12]:
# 1. 조건 만들기
cond = (stroke_df['bmi'].isna()) | (stroke_df['bmi'] < 10) | (stroke_df['bmi']>80)
cond_df = stroke_df[cond]

In [13]:
stroke_df_clean = stroke_df.drop(cond_df.index)

In [14]:
print(len(stroke_df))
print(len(stroke_df_clean))

5110
4907


## 문제 8: 조건 필터링 + 다중 정렬

나이 60세 이상이면서 (고혈압=1 또는 심장질환=1)인 행만 필터링해 생성  
avg_glucose_level 내림차순, 동률이면 bmi 오름차순으로 정렬  
상위 10명에 대해 id, age, avg_glucose_level, bmi, stroke만 출력  

In [16]:
# cond = () & (()|())
cond = (stroke_df['age']>=60) & ((stroke_df['hypertension'] == 1 )|(stroke_df['heart_disease'] == 1 ))
result = stroke_df[cond]

In [17]:
sorted_result = result.sort_values(by=['avg_glucose_level', 'bmi'], ascending=[False, True])

In [18]:
sorted_result[['id', 'age', 'avg_glucose_level', 'bmi', 'stroke']].head(10)

id   age  avg_glucose_level   bmi  stroke
193   69112  68.0             271.74  31.1       1
4983  10995  76.0             267.60  30.5       0
3393  13155  67.0             263.56  26.3       0
4457  50070  62.0             261.67  43.0       0
1067  55862  67.0             254.63  31.0       0
785   17813  69.0             254.60  21.7       0
33    54401  80.0             252.72  30.5       1
386   63732  70.0             251.60  27.1       0
220   20463  81.0             250.89  28.1       1
243   40460  68.0             247.51  40.5       1

## 문제 9: .loc vs .iloc 비교하기
데이터프레임 df에서 age, bmi, stroke 열만 뽑아서 확인해봅시다.
1. .loc를 사용해서 0~2번째 행의 age, bmi, stroke 값을 가져오세요.
2. .iloc를 사용해서 같은 행과 열을 가져오세요.
3. 두 결과가 같은지 눈으로 비교해보세요.

In [7]:
stroke_df

stroke_df.loc[0:2,('age','bmi','stroke')]

age   bmi  stroke
0  67.0  36.6       1
1  61.0   NaN       1
2  80.0  32.5       1

In [11]:
stroke_df.iloc[0:3,[2,9,11]]

age   bmi  stroke
0  67.0  36.6       1
1  61.0   NaN       1
2  80.0  32.5       1